README.md
# NMEA 0183

This haskell implements a [NMEA 0183](https://en.wikipedia.org/wiki/NMEA_0183) parsing library in Haskell.
A small demo project reading the NMEA output of a GPS device can be found [here](https://github.com/marcelbuesing/NMEA-demo).


ChangeLog.md
# Revision history for nmea

## 0.1.0.0  -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


tests/test.hs
{-# LANGUAGE OverloadedStrings #-}

import           Data.Attoparsec.Text
import           Test.Tasty
import           Test.Tasty.HUnit
import           Data.Time.Calendar (fromGregorian)
import           Data.Time.LocalTime (LocalTime(..), ZonedTime(..), midday, midnight, utc, utcToLocalTimeOfDay, zonedTimeZone)

import           NMEA.Common
import           NMEA.GPGGA
import           NMEA.GPRMC

main = defaultMain $ testGroup "NMEA"
  [ testCase "GPGGA" gpggaTest
  , testCase "GPRMC" gprmcTest
  ]


gpggaTest :: Assertion
gpggaTest = do
    let sGpgga = "$GPGGA,120000.000,4250.5589,S,14718.5084,E,1,04,24.4,19.7,M,,,,0000*1F"
        day' = fromGregorian 1970 1 1
        eGpgga = Gpgga {
          _gppgaTimeUTC = ZonedTime (LocalTime day' midday) utc
        , _gppgaLatitude = Latitude {_latitudeValue = 4250.5589, _latitudeDirection = South}
        , _gppgaLongitude = Longitude {_longitudeValue = 14718.5084, _longitudeDirection = East},
          _gppgaGpsQuality = GPSFix,
          _gppgaNumberOfSatellites = 4,
          _gppgaHorizDilutionOfPrecision = 24.4,
          _gppgaAltitude = 19.7,
          _gppgaGeoidalSeparation = 0.0,
          _gppgaAgeDifferentialGPSData = 0.0,
          _gppgaDgpsReferenceStation = DGPSReferenceStation {_dgpsReferenceStation = 0}
          }
    parseOnly gpgga sGpgga @=? Right eGpgga

gprmcTest :: Assertion
gprmcTest = do
    let sGprmc = "$GPRMC,000000.000,A,3723.2475,N,12158.3416,W,0.13,309.62,120598,,*10"
        day' = fromGregorian 1970 1 1
        magVariation = MagneticVariation
          { _magneticVariationValue = Degree {_unDegree = 0.0}
          , _magneticVariationDirection = East
          }
        eGprmc = Gprmc
          { _gprmcTimeUTC = ZonedTime (LocalTime day' midnight) utc
          , _gprmcStatus = DataValid
          , _gprmcLatitude = Latitude {_latitudeValue = 3723.2475, _latitudeDirection = North}
          , _gprmcLongitude = Longitude {_longitudeValue = 12158.3416, _longitudeDirection = West}
          , _gprmcSpeedOverGround = Knot {_unKnot = 0.13}
          , _gprmcCourseOverGround = Degree {_unDegree = 309.62}
          , _gpmrcDate = fromGregorian 1998 5 12
          , _gpmrcMagneticVariation = magVariation
          , _gprmcMode = Autonomous
          }
    parseOnly (gprmc 1900) sGprmc @=? Right eGprmc


src/NMEA/GPRMC.hs
{-# LANGUAGE OverloadedStrings #-}

module NMEA.GPRMC where

import           Control.Applicative ((<|>))
import           Data.Attoparsec.Text
import           Data.Time.Calendar (Day(..))
import           Data.Time.LocalTime (ZonedTime(..))
import           NMEA.Common

data GprmcStatus = DataValid | DataNotValid deriving (Eq, Show)

data GprmcMode = Autonomous | DGPS | DR deriving (Eq, Show)

data MagneticVariation = MagneticVariation
  { _magneticVariationValue     :: Degree
  , _magneticVariationDirection :: LongitudeDirection
  } deriving (Eq, Show)

-- | Recommended minimum specific GNSS data
data Gprmc = Gprmc
  { _gprmcTimeUTC           :: ZonedTime
  , _gprmcStatus            :: GprmcStatus
  , _gprmcLatitude          :: Latitude
  , _gprmcLongitude         :: Longitude
  , _gprmcSpeedOverGround   :: Knot
  , _gprmcCourseOverGround  :: Degree
  , _gpmrcDate              :: Day
  , _gpmrcMagneticVariation :: MagneticVariation
  , _gprmcMode              :: GprmcMode
  } deriving (Eq, Show)

gprmcStatus :: Parser GprmcStatus
gprmcStatus =
      (char 'A' >> return DataValid)
  <|> (char 'V' >> return DataNotValid)

magneticVariation :: Parser MagneticVariation
magneticVariation = do
  v <- degree
  _ <- comma
  d <- longitudeDirection
  return (MagneticVariation v d) <?> "Magnetic Variation"

gprmcMode :: Parser GprmcMode
gprmcMode =
      (char 'A' >> return Autonomous)
  <|> (char 'D' >> return DGPS)
  <|> (char 'E' >> return DR)

gprmc :: Century -> Parser Gprmc
gprmc cen = do
  _       <- string "$GPRMC,"
  time <- timeUTC
  _    <- comma
  stat <- gprmcStatus
  _    <- comma
  lat  <- latitude
  _    <- comma
  lon  <- longitude
  _    <- comma
  spd  <- knot
  _    <- comma
  deg  <- degree
  _    <- comma
  date <- day cen
  _    <- comma
  mv   <- option (MagneticVariation (Degree 0) East) magneticVariation
--  _    <- comma
--  mode <- option Autonomous gprmcMode
  _    <- comma
  _    <- checksum
  return $ Gprmc time stat lat lon spd deg date mv Autonomous


src/NMEA/Common.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE StandaloneDeriving #-}

module NMEA.Common where

import           Control.Applicative ((<|>))
import           Data.Attoparsec.Text
import           Data.Time.Calendar (Day(..), fromGregorian)
import           Data.Time.LocalTime (ZonedTime(..), utc, zonedTimeZone)
import           Data.Time.Format (defaultTimeLocale, formatTime, parseTimeOrError, readSTime)
import           Data.Monoid ((<>))
import qualified Data.Text as T

data LatitudeDirection = North | South deriving (Eq, Show)

data Latitude = Latitude
  { _latitudeValue     :: Double
  , _latitudeDirection :: LatitudeDirection
  } deriving (Eq, Show)

data LongitudeDirection = East | West deriving (Eq, Show)

data Longitude = Longitude
  { _longitudeValue     :: Double
  , _longitudeDirection :: LongitudeDirection
  } deriving (Eq, Show)

-- | Speed in knots, equal to one nautical mile per hour.
newtype Knot  = Knot { _unKnot :: Double } deriving (Eq, Show)

-- | e.g. 90°, max 360°
newtype Degree = Degree { _unDegree :: Double } deriving (Eq, Show)

-- | time as UTC missing date information
timeUTC :: Parser ZonedTime
timeUTC = do
  hh <- T.pack <$> count 2 digit
  mm <- T.pack <$> count 2 digit
  ss <- T.pack <$> count 2 digit
  sss  <- T.pack <$> option "" (char '.' *> many1 digit)
  let ztime = ptime (T.unpack (hh <> mm <> ss <> "." <> sss))
  return $ ztime { zonedTimeZone = utc}
  where ptime s = parseTimeOrError False defaultTimeLocale "%H%M%S%Q" s :: ZonedTime

-- e.g. 2000
type Century = Integer

-- | requires current century to determine correct date e.g. 2000
day :: Century -> Parser Day
day century = do
  dd <- read <$> count 2 digit
  mm <- read <$> count 2 digit
  yy <- read <$> count 2 digit
  return $ fromGregorian (century + yy) mm dd

deriving instance Eq ZonedTime

latitudeDirection :: Parser LatitudeDirection
latitudeDirection =
      (char 'N' >> return North)
  <|> (char 'S' >> return South)

latitude :: Parser Latitude
latitude = do
  v <- double
  _ <- comma
  d <- latitudeDirection
  return (Latitude v d) <?> "Latitude"

longitudeDirection :: Parser LongitudeDirection
longitudeDirection =
      (char 'E' >> return East)
  <|> (char 'W' >> return West)
  <?> "Longitude Direction"

longitude :: Parser Longitude
longitude = do
  v <- double
  _ <- comma
  d <- longitudeDirection
  return (Longitude v d) <?> "Longitude"

knot :: Parser Knot
knot = Knot <$> double

degree :: Parser Degree
degree = Degree <$> double

checksum :: Parser Int
checksum =  char '*' *> hexadecimal :: Parser Int

comma :: Parser Char
comma = char ','


src/NMEA/GPGGA.hs
{-# LANGUAGE OverloadedStrings #-}

module NMEA.GPGGA where

import           Control.Applicative ((<|>))
import           Data.Attoparsec.Text
import           Data.Monoid ((<>))
import qualified Data.Text as T
import           Data.Time.LocalTime (ZonedTime(..))

import NMEA.Common

-- $GPGGA,HHMMSS.ss,BBBB.BBBB,b,LLLLL.LLLL,l,Q,NN,D.D,H.H,h,G.G,g,A.A,RRRR*PP

-- $GPGGA,235947.000,0000.0000,N,00000.0000,E,0,00,0.0,0.0,M,,,,0000*00
-- $GPGGA,092204.999,4250.5589,S,14718.5084,E,1,04,24.4,19.7,M,,,,0000*1F

-- | GPS quality indicator
data PositionFixIndicator =
    InvalidFix
  | GPSFix
  | DGPSFix
  | DeadReckoningMode deriving (Eq, Show)

newtype DGPSReferenceStation = DGPSReferenceStation { _dgpsReferenceStation :: Int } deriving (Eq, Show)

-- | Global Positioning System Fix Data
data Gpgga = Gpgga
  { _gppgaTimeUTC                  :: ZonedTime
  , _gppgaLatitude                 :: Latitude
  , _gppgaLongitude                :: Longitude
  , _gppgaGpsQuality               :: PositionFixIndicator
  , _gppgaNumberOfSatellites       :: Int
  , _gppgaHorizDilutionOfPrecision :: Double
  , _gppgaAltitude                 :: Double
  , _gppgaGeoidalSeparation        :: Double
  , _gppgaAgeDifferentialGPSData   :: Double
  , _gppgaDgpsReferenceStation     :: DGPSReferenceStation
  } deriving (Eq, Show)

positionFixIndicator :: Parser PositionFixIndicator
positionFixIndicator =
      (char '0' >> return InvalidFix)
  <|> (char '1' >> return GPSFix)
  <|> (char '2' >> return DGPSFix)
  <|> (char '6' >> return DeadReckoningMode)
  <?> "GPS Quality"

numberOfSatellites :: Parser Int
numberOfSatellites = read <$> count 2 digit <?> "Number of satellites"

dgpsReferenceStation :: Parser DGPSReferenceStation
dgpsReferenceStation = do
  d <- read <$> count 4 digit
  return (DGPSReferenceStation d) <?> "DGPSReferenceStation"

gpgga :: Parser Gpgga
gpgga = do
  string "$GPGGA,"
  time <- timeUTC
  _    <- comma
  lat  <- latitude
  _    <- comma
  lon  <- longitude
  _    <- comma
  qual <- positionFixIndicator
  _    <- comma
  nsat <- decimal :: Parser Int
  _    <- comma
  dilu <- double
  _    <- comma
  alti <- double
  _    <- comma *> char 'M' *> ","
  geoi <- option 0 double
  _    <- comma *> option 'M' (char 'M') *> comma
  age  <- option 0 double
  _    <- comma
  dgps <- option (DGPSReferenceStation 0) dgpsReferenceStation
  _    <- checksum
  return (Gpgga time lat lon qual nsat dilu alti geoi age dgps) <?> "GPPGA"
